### PR TITLE
VMware: vCenter plugin : new module

### PR DESCRIPTION
**Pull request by CastawayEGR**
##### SUMMARY
Add new module for managing vCenter Extensions.

##### ISSUE TYPE

 - New Module Pull Request

##### ANSIBLE VERSION

```
2.7-devel
```

### CODE REVIEW

File: lib/ansible/modules/cloud/vmware/vcenter_plugin.py [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+  email:
+    description:
+    - Required for C(state=present). Administrator Email to use for Plugin.
+  url:
+    description:
+    - Required for C(state=present). Link to server hosting plugin zip file to install.
+  ssl_thumbprint:
+    description:
+    - Required for C(state=present). SSL thumbprint of the plugin hosting server.
+  type:
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vcenter_plugin.py]: Could you please rename this parameter to something more descriptive ?

File: lib/ansible/modules/cloud/vmware/vcenter_plugin.py [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+  state:
+    description:
+    - Add or remove vCenter Plugin.
+    choices: [absent, present]
+    default: present
+extends_documentation_fragment: vmware.documentation
+'''
+
+EXAMPLES = '''
+- name: vCenter Plugin Examples
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vcenter_plugin.py]: Line 68 to 73 not required.

File: lib/ansible/modules/cloud/vmware/vcenter_plugin.py [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+
+try:
+    import datetime
+    HAS_DT = True
+except ImportError:
+    HAS_DT = False
+
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.vmware import (HAS_PYVMOMI, connect_to_api, vmware_argument_spec)
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vcenter_plugin.py]: Use 

```python
from ansible.module_utils.vmware import PyVmomi, ...
````

File: lib/ansible/modules/cloud/vmware/vcenter_plugin.py [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+        name=dict(type='str', required=False),
+        url=dict(type='str', required=False),
+        ssl_thumbprint=dict(type='str', required=False),
+        type=dict(type='str', default='vsphere-client-serenity', required=False),
+        state=dict(type='str', default='present', choices=['absent', 'present']),
+    ))
+
+    module = AnsibleModule(argument_spec=argument_spec, supports_check_mode=False)
+
+    if not HAS_PYVMOMI:
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vcenter_plugin.py]: not required if `PyVmomi` is used.

File: lib/ansible/modules/cloud/vmware/vcenter_plugin.py [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+    email = module.params['email']
+    desc = module.params['description']
+    name = module.params['name']
+    company = module.params['company']
+    type = module.params['type']
+    url = module.params['url']
+    thumbprint = module.params['ssl_thumbprint']
+
+    try:
+        content = connect_to_api(module, False)
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vcenter_plugin.py]: not required if `PyVmomi` is used.

File: lib/ansible/modules/cloud/vmware/vcenter_plugin.py [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+        em = content.extensionManager
+        key_check = em.FindExtension(extension_key)
+        results = dict(changed=False, result=dict())
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vcenter_plugin.py]: rename return dict `result` to more specific name.

**Review by Akasurde — CHANGES_REQUESTED** [review]
Could you please check if we can add unit tests for this change ?

**Comment by CastawayEGR** [comment]
To the best of my knowledge it does not look like unit tests can be added.

**Review by CastawayEGR — COMMENTED** [review]
@Akasurde recommended changes have been implemented at this time.

**Comment by CastawayEGR** [comment]
@Akasurde can you advise on the next steps to remove the needs_revision label?

**Comment by Akasurde** [comment]
@CastawayEGR Thanks for updating code as per code review comments. I will test it and let you know, that will hopefully remove needs_revision tag. Thanks.

File: lib/ansible/modules/cloud/vmware/vcenter_plugin.py [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+RETURN = """
+result:
+    description: information about performed operation
+    returned: always
+    type: string
+    sample: "'com.acme.plugin' installed."
+"""
+
+try:
+    from pyVmomi import vim, vmodl
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vcenter_plugin.py]: vmodl is not used, so can be safely removed.

File: lib/ansible/modules/cloud/vmware/vcenter_plugin.py [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+except ImportError:
+    pass
+
+try:
+    import datetime
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vcenter_plugin.py]: I think datetime is basic datatype of Python and is available in all installation. So this check will not be required.

**Comment by Akasurde** [comment]
@CastawayEGR I am sorry, but can we rename this to `vcenter_extension` or something similar. I found that it is more apt as we are dealing extension than plugins. People might get confused by name, as extension is more common than plugins. Let me know your views. Thanks.

**Comment by CastawayEGR** [comment]
@Akasurde no worries at all. I had it named this way at first but thought it might be confusing. I agree with you considering how it is called in the pyvmomi documentation. All recommended edits have been made at this time. Thanks!

**Comment by CastawayEGR** [comment]
@Akasurde is there anything else you need me to do on this end to get this module added?

**Review by Akasurde — APPROVED** [review]

**Comment by Akasurde** [comment]
rebuild_merge

**Comment by Akasurde** [comment]
@CastawayEGR Thanks for the contribution.

**Pull request merged**